src/proposed/0140-budget-limit-threshold.md

```diff
@@ -0,0 +1,469 @@
+<!--
+  SPDX-License-Identifier: CC-BY-SA-4.0
+  Copyright 2023 UNSW
+-->
+
+# MCS: Budget limit thresholds on endpoints for SC Donation
+
+- Author: Mitchell Johnston
+- Proposed: 2023-08-15
+
+## Summary
+
+This is a two stage RFC that proposes to add a new `budget threshold' to
+endpoints, which would only allow scheduling context donation to occur if the
+SC's available budget exceeds the threshold. This would provide a mechanism to
+prevent budget-expiry from occurring in passive servers.
+
+As a second optional stage, this threshold value could also restrict the budget
+that a passive server is permitted to consume on a donated SC. We refer to this
+as a `budget limit'. This would help bound priority inversion in a system as
+well as reducing the trust that clients must place in passive servers they call.
+
+## Motivation
+
+### Thresholds
+
+This change regarding thresholds is motivated by a desire to prevent budget
+expiry from occurring inside passive servers, improving worst-case performance
+and response time analysis. Additionally, it would shift timeout exceptions into
+being a true error case, rather than an unavoidable, but routine occurrence on
+many systems. This is likely to simplify system design.
+
+Currently, there are almost no restrictions on the amount of budget required to
+enter a passive server. This almost inevitably leads to budget expiry, resulting
+in undesirable longer blocking times for other clients of the passive server.
+This must also be accounted for in schedulability analysis of a system and is
+particularly important for servers with clients of varying criticality.
+
+The kernel currently supports a mechanism to manage budget expiry, timeout
+handlers. These allow an error-handling thread to intervene and resolve the
+budget expiry. However, as they address the issue after it occurs, these have
+some limitations. Timeout handlers can provide extra budget, however, it is not
+preferable to reward clients with additional budget just because they are
+blocking the rest of the system. Alternatively, the server can be reverted to a
+previous state, however this wastes any work the server performed and can be
+expensive or impractical, depending on the nature of the passive server.
+
+This RFC proposes a new mechanism, endpoint thresholds, which aim to prevent
+budget expiry from occurring in passive servers, rather than reacting to it
+after it occurs.
```

Comment:
The motivation here would be improved greatly by adding that this is additional, demostrating why it's needed, without saying all the other options are bad. Reacting is a legitiamte way to build a system, especially if your WCET is very loose. They have legitimate use cases depending on client policy.